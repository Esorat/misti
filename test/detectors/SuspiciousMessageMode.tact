// Define custom constants
const CustomSendRemainingValue: Int = 64;
const CustomSendAllBalance: Int = 128;
const CustomSendIgnoreErrors: Int = 1;

contract SendParametersTestContract {
    // Correct usage: should not trigger any warnings
    fun correctUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: CustomSendRemainingValue
        });
    }

    // Duplicate flag usage: should trigger a warning about flags used multiple times
    fun duplicateFlagUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: CustomSendRemainingValue | CustomSendRemainingValue
        });
    }

    // Invalid operator usage: should trigger a warning about using '+' instead of '|'
    fun invalidOperatorUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: CustomSendRemainingValue + CustomSendAllBalance
        });
    }

    // Integer literal usage: should trigger a warning about using integer literals
    fun integerLiteralUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: 64 // Integer literal instead of symbolic constant
        });
    }

    // Correct combination usage: should not trigger any warnings
    fun correctCombinationUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: CustomSendRemainingValue | CustomSendAllBalance | CustomSendIgnoreErrors
        });
    }

    // Multiple issues: should trigger multiple warnings
    fun multipleIssues() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: CustomSendRemainingValue + CustomSendRemainingValue + 64 // Duplicate flags, '+' operator, integer literal
        });
    }

    // Complex expression with nested invalid operators
    fun complexInvalidOperator() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: (CustomSendRemainingValue * CustomSendAllBalance) - CustomSendIgnoreErrors
        });
    }

    // Define functionReturningLiteral within the contract scope
    fun functionReturningLiteral(): Int {
        return 64;
    }

    // Using a function call that returns an integer literal: should trigger a warning
    fun functionCallWithLiteral() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: self.functionReturningLiteral() // Uses integer literal instead of symbolic constant
        });
    }

// Using undefined symbolic constants: should trigger a warning if the constant is not recognized
 fun undefinedSymbolicConstant() {
    send(SendParameters{
        to: sender(),
        value: 0,
        mode: CustomSendIgnoreErrors | CustomSendAllBalance
    });
}

    // Correct usage with variables
    fun correctUsageWithVariables() {
        let modeFlag: Int = CustomSendRemainingValue | CustomSendIgnoreErrors;
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: modeFlag
        });
    }

    // Invalid usage with variables containing integer literals
    fun invalidUsageWithVariableLiteral() {
        let modeFlag: Int = 64;
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: modeFlag // modeFlag contains an integer literal
        });
    }
}
