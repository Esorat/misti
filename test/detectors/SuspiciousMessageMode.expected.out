[MEDIUM] SuspiciousMessageMode: Flag 'CustomSendRemainingValue' is used multiple times in mode expression
test/detectors/SuspiciousMessageMode.tact:21:46:
  20 |             value: 0,
> 21 |             mode: CustomSendRemainingValue | CustomSendRemainingValue
                                                    ^
  22 |         });
Help: Use each flag at most once in the mode expression
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:30:19:
  29 |             value: 0,
> 30 |             mode: CustomSendRemainingValue + CustomSendAllBalance
                         ^
  31 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Integer literals should not be used in mode expression; use symbolic constants instead
test/detectors/SuspiciousMessageMode.tact:39:19:
  38 |             value: 0,
> 39 |             mode: 64 // Integer literal instead of symbolic constant
                         ^
  40 |         });
Help: Replace integer literals with symbolic flag constants
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:57:19:
  56 |             value: 0,
> 57 |             mode: CustomSendRemainingValue + CustomSendRemainingValue + 64 // Duplicate flags, '+' operator, integer literal
                         ^
  58 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Flag 'CustomSendRemainingValue' is used multiple times in mode expression
test/detectors/SuspiciousMessageMode.tact:57:46:
  56 |             value: 0,
> 57 |             mode: CustomSendRemainingValue + CustomSendRemainingValue + 64 // Duplicate flags, '+' operator, integer literal
                                                    ^
  58 |         });
Help: Use each flag at most once in the mode expression
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Integer literals should not be used in mode expression; use symbolic constants instead
test/detectors/SuspiciousMessageMode.tact:57:73:
  56 |             value: 0,
> 57 |             mode: CustomSendRemainingValue + CustomSendRemainingValue + 64 // Duplicate flags, '+' operator, integer literal
                                                                               ^
  58 |         });
Help: Replace integer literals with symbolic flag constants
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:66:19:
  65 |             value: 0,
> 66 |             mode: (CustomSendRemainingValue * CustomSendAllBalance) - CustomSendIgnoreErrors
                         ^
  67 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:66:20:
  65 |             value: 0,
> 66 |             mode: (CustomSendRemainingValue * CustomSendAllBalance) - CustomSendIgnoreErrors
                          ^
  67 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode